src/compiler/checker.ts

```diff
@@ -14441,11 +14442,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getResolvedApparentTypeOfMappedType(type: MappedType) {
-        const typeVariable = getHomomorphicTypeVariable(type);
-        if (typeVariable && !type.declaration.nameType) {
-            const constraint = getConstraintOfTypeParameter(typeVariable);
+        if (contains(homomorphicMappedTypeApparentTypeStack, type)) {
+            return type;
+        }
+        const mappedType = type.target as MappedType || type;
+        const typeVariable = getHomomorphicTypeVariable(mappedType);
+        if (typeVariable && !mappedType.declaration.nameType) {
+            let constraint: Type | undefined;
+            if (!type.target) {
+                constraint = getConstraintOfTypeParameter(typeVariable);

```

Comment:
The other branch is now using `getApparentType`. If that's the correct thing to do then I feel like this probably should call that too. This would fix issues like this one:
```ts
declare function func2<A extends readonly any[], B extends A>(
  arr: A,
  fields: [
    // A rest element type must be an array type.(2574)
    ...{
      [K in keyof B]: { v: B[K] };
    },
  ],
): [A, B];

declare const tuple: [number | string, number | boolean, number | bigint];

const result = func2(tuple, [{ v: "foo" }, { v: true }, { v: 6 }]);
```

Reply:
You probably could... or you could move the apparent type check forward and replace `everyType(constraint, isArrayOrTupleType)` with something like `everyType(constraint, isApparentArrayOrTupleType)`, provided it's not critical the instantiation here be with the apparent types (and I don't *think* the instantiation using an apparent constraint is the important part, since mapped types have lazy apparent type members). It's possible changing the should-this-be-a-mapped-array check we use *everywhere* to something that looks into the apparent types of the members of the constraint could be prudent.